### PR TITLE
CodeMirror: make Cmd-f trigger search globally

### PR DESCRIPTION
**Pull request by oleggromov**
## Description
As we're approaching shipping CodeMirror-based blob view for everyone, we need to take care of Cmd/Ctrl-F behaviour.
This diff adds it:
- as a global keyboard shortcut to `Layout` - theoretically, it should trigger search everywhere in the app, when the focus isn't hijacked and code mirror blob view is visible
- to the SearchInput, as it has it's own hotkey handling - and we're triggering the CodeMirror search instead

Whenever search triggers, I am actually pressing `Cmd-F`:

https://user-images.githubusercontent.com/2196347/186660156-43a7e00a-5c2b-4ec8-8d9c-f35e61498abf.mov

## Test plan

0. Run the app locally 
1. Set `enableCodeMirrorFileView` to `true` in settings
2. Check that with the focus inside the codemirror view, inside the search bar, or anywhere on the page Cmd-F triggers blob view search
3. Check that outside the page with the blob view, it triggers browser's native search (unless hijacked by some other component)
4. Toggle the setting off and check that the search isn't broken

## App preview:

- [Web](https://sg-web-og-cm-trigger-search.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-fkcmdfctag.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

File: client/web/src/Layout.tsx [hunk 1]
```diff
@@ -238,121 +246,124 @@ export const Layout: React.FunctionComponent<React.PropsWithChildren<LayoutProps
...
-                            : isCommunitySearchContextPage
-                            ? 'low-profile-with-logo'
-                            : 'default'
-                    }
-                    minimalNavLinks={minimalNavLinks}
-                    isSearchAutoFocusRequired={!isSearchAutoFocusRequired}
-                    isRepositoryRelatedPage={isRepositoryRelatedPage}
-                    showKeyboardShortcutsHelp={showKeyboardShortcutsHelp}
-                    onHandleFuzzyFinder={setIsFuzzyFinderVisible}
+            <CodeMirrorContext.Provider value={{ setCodeMirrorBlobEditor, triggerSearchIfAvailable }}>
```
Comment (olafurpg) [inline client/web/src/Layout.tsx:249]: Just curious: was it necessary to wrap all these components under `CodeMirrorContext.Provider` or would it have been possible to only wrap the components that use `setCodeMirrorBlobEditor` and `triggerSearchIfAvailable`?

**Comment by oleggromov** [comment]
@sourcegraph/frontend-platform I tagged you because I touched the `Layout.tsx` and you may want to be aware of this change.

File: client/search-ui/src/input/CodeMirrorQueryInput.tsx [hunk 2]
```diff
@@ -469,6 +478,17 @@ const [callbacksField, setCallbacks] = createUpdateableField<
                 return false
             },
         },
+        {
+            key: 'Mod-f',
+            run: view => {
+                const { onTriggerSearch } = view.state.field(callbacks)
+                if (onTriggerSearch) {
+                    onTriggerSearch()
```
Comment (olafurpg) [inline client/search-ui/src/input/CodeMirrorQueryInput.tsx:486]: This case seems to be triggering with the old blob view, preventing Cmd+F to work. We should only return `true` if the new blob view is visible

**Review by olafurpg — CHANGES_REQUESTED** [review]

File: client/web/src/Layout.tsx (hunk 1 above)
Comment (oleggromov) [reply client/web/src/Layout.tsx:249]: Strictly speaking it isn't necessary but I much prefer having a context around it (usual thing in React apps, we've got multiple contexts ourselves) than passing props down through tens of different components (which we also do to be fair).

File: client/search-ui/src/input/CodeMirrorQueryInput.tsx (hunk 2 above)
Comment (oleggromov) [reply client/search-ui/src/input/CodeMirrorQueryInput.tsx:486]: Great catch, thank you! Will fix now!

File: client/search-ui/src/input/LazyMonacoQueryInput.tsx [hunk 3]
```diff
@@ -4,6 +4,7 @@ import classNames from 'classnames'
 
 import { SettingsExperimentalFeatures } from '@sourcegraph/shared/src/schema/settings.schema'
 import { lazyComponent } from '@sourcegraph/shared/src/util/lazyComponent'
+import { useCodeMirrorContext } from '@sourcegraph/web/src/repo'
```
Comment (valerybugakov) [inline client/search-ui/src/input/LazyMonacoQueryInput.tsx:7]: The `search-ui` package was extracted from the web package to be able to share logic between the VSCode extension and the web application. We should not have imports from the `@sourcegraph/web` package here. 

The functionality required for the component should be moved to the `search-ui` package if it's not specific to the web application, or it should be injected into components here via props or React context. Let me know if it doesn't make sense.

**Review by valerybugakov — CHANGES_REQUESTED** [review]
The bundlesize check [shows](https://github.com/sourcegraph/sourcegraph/pull/40792/checks?check_run_id=8016350532) a massive increase in the main chunk bundle. Based on the functionality added in this PR, it should not be happening. It might be related to the new import from the `web` package into the `search-ui` package. 

The best way to investigate this issue would be using the `analyze-bundle` NPM script defined in the `web` package. It uses [the statoscope](https://github.com/statoscope/statoscope) Webpack plugin under the hood, which should help find the root cause.

File: client/search-ui/src/input/LazyMonacoQueryInput.tsx (hunk 3 above)
Comment (oleggromov) [reply client/search-ui/src/input/LazyMonacoQueryInput.tsx]: Thank you, this makes sense. I've turned to passing the callback through props as you recommended.

**Review by valerybugakov — COMMENTED** [review]
@oleggromov, did you have a chance to look into bundlesize increase of the `main` chunk? We need to understand where it's coming from and either prevent it from happening or increase the main chunk limit (a less desirable outcome).

**Comment by oleggromov** [comment]
> @oleggromov, did you have a chance to look into bundlesize increase of the `main` chunk? We need to understand where it's coming from and either prevent it from happening or increase the main chunk limit (a less desirable outcome).

I am trying to figure this out now. The import from `web/` is removed and I am not sure where it comes from.
Will post an update soon.

**Comment by oleggromov** [comment]
I think I've found out what the issue is:

1. To make `Cmd+F` trigger CodeMirror search from _everywhere_ in the app, I import `useCodeMirrorBlobSearch` into `Layout`
2. `useCodeMirrorBlobSearch` imports `openSearchPanel` from `@codemirror/search`, which imports the [entire CodeMirror](https://github.com/codemirror/search/blob/main/package.json#L29-L30) as it seems
3. This makes the entire CodeMirror go into the main chunk, which increases the bundle size

Not sure I see an easy way around this right now, especially without changing the architecture of key handling.

One (kinda crazy) idea is to disconnect `Cmd+F` handling (`Shortcut` component basically) and CodeMirror response to it (`openSearchCallback` invocation) but this will require a big refactoring and a non-obvious architecture.

@sourcegraph/frontend-platform maybe you have other ideas?

**A more radical solution** would be to say we're definitely getting rid of Monaco, which should reduce bundle size, and just proceed with shipping it as is.

But for this, I need @olafurpg and @macraig input at least.

**Comment by macraig** [comment]
@oleggromov what would the drawbacks and/or consequences of getting rid of Monaco be?

**Comment by oleggromov** [comment]
> @oleggromov what would the drawbacks and/or consequences of getting rid of Monaco be?

I am guessing it is gonna happen anyways in the future as we migrate to CodeMirror. But I am not sure what the current plan is.

**Comment by olafurpg** [comment]
The timeline for Monaco deprecation might be long since Monaco is used in various places (several admin UIs, and batch changes). I don't think we can realistically remove it from the app within the short period I'd like to get Cmd+F working globally with the new blob view. I'll take over this issue and investigate alternative solutions once the extension migration is done. Maybe Felix Kling has some creative ideas once he returns next week. In the meantime, I recommend we close this PR for now and leave the Cmd+F issue open.

**Comment by oleggromov** [comment]
This makes sense to me!
Thank you @olafurpg

**Pull request closed without merging**